File: Vision2language.StoryGenerator/Api/StoryLocationBase.cs
using SyntheticVideo2language.Story.Data;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SyntheticVideo2language.Story.Api
{
    public abstract class StoryLocationBase : IStoryItem
    {
        public StoryLocationBase()
        {
            this.PossibleActions = new List<StoryActionBase>();
        }
        public abstract string Description { get; set; }
        public eStoryItemType StoryItemType => eStoryItemType.Location;
        public abstract List<StoryActionBase> PossibleActions { get; set; }
        public abstract StoryActionBase GetNextRandomValidAction();
    }
}

File: api/StoryLocationBase.lua
StoryLocationBase = class(IStoryItem, function(o, description, possibleActions)
    IStoryItem.init(o, description, eStoryItemType.Location)
    o.PossibleActions = possibleActions
    o.LocationId = Guid().Id
end
)

function StoryLocationBase:__tostring()
    return self.Description
  end

function StoryLocationBase:__eq(other)
    return other and other:is_a(StoryLocationBase) and self.LocationId == other.LocationId
end